<a href="https://colab.research.google.com/github/zayedfarooq/Unified-Mentor-Projects/blob/main/fifa_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
#import necessary libraries
import pandas as pd
import plotly.express as px
import plotly.figure_factory as ff

In [4]:
#importing the necessary files for analysis
df1 = pd.read_csv("/content/WorldCupMatches.csv")
df2 = pd.read_csv("/content/WorldCupPlayers.csv")
df3 = pd.read_csv("/content/WorldCups.csv")

In [5]:
#top 5 values
df1.head(5)

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

In [6]:
#bottom 5 values
df1.tail(5)

Year Datetime Stage Stadium City Home Team Name  Home Team Goals  \
4567   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4568   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4569   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4570   NaN      NaN   NaN     NaN  NaN            NaN              NaN   
4571   NaN      NaN   NaN     NaN  NaN            NaN              NaN   

      Away Team Goals Away Team Name Win conditions  Attendance  \
4567              NaN            NaN            NaN         NaN   
4568              NaN            NaN            NaN         NaN   
4569              NaN            NaN            NaN         NaN   
4570              NaN            NaN            NaN         NaN   
4571              NaN            NaN            NaN         NaN   

      Half-time Home Goals  Half-time Away Goals Referee Assistant 1  \
4567                   NaN                   NaN     NaN         NaN   
4568                   NaN                   NaN     NaN         NaN   
4569                   NaN                   NaN     NaN         NaN   
4570                   NaN                   NaN     NaN         NaN   
4571                   NaN                   NaN     NaN         NaN   

     Assistant 2  RoundID  MatchID Home Team Initials Away Team Initials  
4567         NaN      NaN      NaN                NaN                NaN  
4568         NaN      NaN      NaN                NaN                NaN  
4569         NaN      NaN      NaN                NaN                NaN  
4570         NaN      NaN      NaN                NaN                NaN  
4571         NaN      NaN      NaN                NaN                NaN

In [7]:
#Checking for null values
df1.isnull().sum()

Year                    3720
Datetime                3720
Stage                   3720
Stadium                 3720
City                    3720
Home Team Name          3720
Home Team Goals         3720
Away Team Goals         3720
Away Team Name          3720
Win conditions          3720
Attendance              3722
Half-time Home Goals    3720
Half-time Away Goals    3720
Referee                 3720
Assistant 1             3720
Assistant 2             3720
RoundID                 3720
MatchID                 3720
Home Team Initials      3720
Away Team Initials      3720
dtype: int64

In [8]:
#Dropping the null values from dataframe
df1_cleaned = df1.dropna()

In [9]:
df1_cleaned.isnull().sum()

Year                    0
Datetime                0
Stage                   0
Stadium                 0
City                    0
Home Team Name          0
Home Team Goals         0
Away Team Goals         0
Away Team Name          0
Win conditions          0
Attendance              0
Half-time Home Goals    0
Half-time Away Goals    0
Referee                 0
Assistant 1             0
Assistant 2             0
RoundID                 0
MatchID                 0
Home Team Initials      0
Away Team Initials      0
dtype: int64

In [10]:
#Checking for duplicated values and dropping/removing them
df1_cleaned.duplicated().sum()
df1_without_duplicates = df1_cleaned.drop_duplicates()

In [11]:
df1_without_duplicates.duplicated().sum()

0

In [12]:
#top 5 values
df2.head(5)

RoundID  MatchID Team Initials           Coach Name Line-up  Shirt Number  \
0      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
1      201     1096           MEX     LUQUE Juan (MEX)       S             0   
2      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   
3      201     1096           MEX     LUQUE Juan (MEX)       S             0   
4      201     1096           FRA  CAUDRON Raoul (FRA)       S             0   

        Player Name Position Event  
0       Alex THEPOT       GK   NaN  
1   Oscar BONFIGLIO       GK   NaN  
2  Marcel LANGILLER      NaN  G40'  
3      Juan CARRENO      NaN  G70'  
4   Ernest LIBERATI      NaN   NaN

In [13]:
#bottom 5 values
df2.tail(5)

RoundID    MatchID Team Initials               Coach Name Line-up  \
37779   255959  300186501           ARG  SABELLA Alejandro (ARG)       N   
37780   255959  300186501           GER       LOEW Joachim (GER)       N   
37781   255959  300186501           ARG  SABELLA Alejandro (ARG)       N   
37782   255959  300186501           GER       LOEW Joachim (GER)       N   
37783   255959  300186501           ARG  SABELLA Alejandro (ARG)       N   

       Shirt Number Player Name Position       Event  
37779            19     ALVAREZ      NaN         NaN  
37780             6     KHEDIRA      NaN         NaN  
37781            20      AGUERO      NaN  IH46' Y65'  
37782            21     MUSTAFI      NaN         NaN  
37783            23     BASANTA      NaN         NaN

In [14]:
#Checking for null values
df2.isnull().sum()

RoundID              0
MatchID              0
Team Initials        0
Coach Name           0
Line-up              0
Shirt Number         0
Player Name          0
Position         33641
Event            28715
dtype: int64

In [15]:
#Checking for null values
df3.isnull().sum()

Year              0
Country           0
Winner            0
Runners-Up        0
Third             0
Fourth            0
GoalsScored       0
QualifiedTeams    0
MatchesPlayed     0
Attendance        0
dtype: int64

In [16]:
#Checking for duplicate values
df2.duplicated().sum()

736

In [17]:
#Dropping the duplicated values
df2_without_duplicates = df2.drop_duplicates()

In [18]:
df2_without_duplicates.duplicated().sum()

0

In [19]:
#temporary check
data_Messi1 = df2_without_duplicates[df2_without_duplicates['Player Name']=='MESSI']
data_Messi1.count()

RoundID          17
MatchID          17
Team Initials    17
Coach Name       17
Line-up          17
Shirt Number     17
Player Name      17
Position          8
Event             6
dtype: int64

In [20]:
#temporary check
data_Messi = df2[df2['Player Name']=='MESSI']
data_Messi.count()

RoundID          21
MatchID          21
Team Initials    21
Coach Name       21
Line-up          21
Shirt Number     21
Player Name      21
Position         12
Event             6
dtype: int64

In [21]:
#temporary check
data_Ronaldo1 = df2_without_duplicates[df2_without_duplicates['Player Name']=='RONALDO']
data_Ronaldo1.count()

RoundID          33
MatchID          33
Team Initials    33
Coach Name       33
Line-up          33
Shirt Number     33
Player Name      33
Position          7
Event            18
dtype: int64

In [22]:
#temporary check
data_Ronaldo = df2[df2['Player Name']=='RONALDO']
data_Ronaldo.count()

RoundID          33
MatchID          33
Team Initials    33
Coach Name       33
Line-up          33
Shirt Number     33
Player Name      33
Position          7
Event            18
dtype: int64

In [23]:
#top 5 values
df3.head(5)

Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0           70              13             18    590.549  
1           70              16             17    363.000  
2           84              15             18    375.700  
3           88              13             22  1.045.246  
4          140              16             26    768.607

In [24]:
#bottom 5 values
df3.tail(5)

Year       Country   Winner   Runners-Up        Third          Fourth  \
15  1998        France   France       Brazil      Croatia     Netherlands   
16  2002   Korea/Japan   Brazil      Germany       Turkey  Korea Republic   
17  2006       Germany    Italy       France      Germany        Portugal   
18  2010  South Africa    Spain  Netherlands      Germany         Uruguay   
19  2014        Brazil  Germany    Argentina  Netherlands          Brazil   

    GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
15          171              32             64  2.785.100  
16          161              32             64  2.705.197  
17          147              32             64  3.359.439  
18          145              32             64  3.178.856  
19          171              32             64  3.386.810

In [25]:
#top 5 values
df1_cleaned.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

In [26]:
#bottom 5 values
df1_cleaned.tail(5)

Year              Datetime                     Stage  \
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
847     Arena Fonte Nova        Salvador     Netherlands              0.0   
848     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
849   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   

     Away Team Goals Away Team Name                         Win conditions  \
847              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
848              7.0        Germany                                          
849              0.0      Argentina    Argentina win on penalties (2 - 4)    
850              3.0    Netherlands                                          
851              0.0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
847     51179.0                   0.0                   0.0   
848     58141.0                   0.0                   5.0   
849     63267.0                   0.0                   0.0   
850     68034.0                   0.0                   2.0   
851     74738.0                   0.0                   0.0   

                   Referee                  Assistant 1  \
847  Ravshan IRMATOV (UZB)  RASULOV Abduxamidullo (UZB)   
848  RODRIGUEZ Marco (MEX)      TORRENTERA Marvin (MEX)   
849     C�neyt �AKIR (TUR)         DURAN Bahattin (TUR)   
850  HAIMOUDI Djamel (ALG)         ACHIK Redouane (MAR)   
851   Nicola RIZZOLI (ITA)        Renato FAVERANI (ITA)   

                  Assistant 2   RoundID      MatchID Home Team Initials  \
847  KOCHKAROV Bakhadyr (KGZ)  255953.0  300186488.0                NED   
848     QUINTERO Marcos (MEX)  255955.0  300186474.0                BRA   
849         ONGUN Tarik (TUR)  255955.0  300186490.0                NED   
850   ETCHIALI Abdelhak (ALG)  255957.0  300186502.0                BRA   
851      Andrea STEFANI (ITA)  255959.0  300186501.0                GER   

    Away Team Initials  
847                CRC  
848                GER  
849                ARG  
850                NED  
851                ARG

In [27]:
#Displaying a line plot of Total Matches And Qualified Teams over the Years
fig=px.line(df3,x="Year",y=["MatchesPlayed","QualifiedTeams"],color_discrete_map={"MatchesPlayed":"orange","QualifiedTeams":"green"}
        ,labels={"value":"Total"})
fig.update_layout(title="Total Qualified Teams and Matches Played Over The Years")

In [28]:
#Displaying a Line Plot of Goals Score Over the years at the World Cup
fig=px.line(df3,x="Year",y="GoalsScored",text="GoalsScored")
fig.update_layout(title="Goals Scored Over The Years")
fig.update_traces(textposition="bottom center")
fig.show()

--**Insights from the above graphs**--


The above graphs show an increase in all the three features that is Matches played, Qualified teams and the Goals scored over the different years at the FIFA World cup. So in Summary we can say that more the teams qualified at the world cup would result in more the matches being played and inturn result in more the goals being scored

In [29]:
#top 5 values
df3.head()

Year      Country      Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay     Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy       Italy  Czechoslovakia  Germany     Austria   
2  1938       France       Italy         Hungary   Brazil      Sweden   
3  1950       Brazil     Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany FR         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0           70              13             18    590.549  
1           70              16             17    363.000  
2           84              15             18    375.700  
3           88              13             22  1.045.246  
4          140              16             26    768.607

In [30]:
#Using groupby Function
avgAttendance=df1_without_duplicates.groupby("Year")["Attendance"].mean().reset_index()

In [31]:
# displaying a bar graph using plotly of average attendance over the years at the world cup
px.bar(avgAttendance,x="Year",y="Attendance",text="Year",color="Attendance",title="Average Attandance Over The World Cups",
       labels={"Attendance" : "Average Attendance"})

In [32]:
#Using grouby function
totalAttendance=df1_without_duplicates.groupby("Year")["Attendance"].sum().reset_index()

In [33]:
# displaying a bar graph using plotly of total attendance over the years at the world cup
px.bar(totalAttendance,x="Year",y="Attendance",text="Year",color="Attendance",title="Total Attandance Over The World Cups",
       labels={"Attendance" : "Total Attendance"})

--**Insights from the above graphs**--


The above graphs show the average and total attendance over the years at the world cup and from the both graphs we can see that there has been a massive increase in attendance numbers at the World cup with 1994 leading the attendance numbers followed by 2014 and 2016 respectively.. As we know that 1994 World Cup was the first world cup with 32 teams showing that with the increasing number of teams we can expect an increase in the attendance numbers as well.

In [34]:
#Renaming the team names to Germany as they repesent a single nation now
correct={'German DR':'Germany','Germany FR':'Germany'}

In [35]:
#Replacing the names in dataset
df3["Winner"]=df3["Winner"].replace(correct)
df3["Runners-Up"]=df3["Runners-Up"].replace(correct)
df3["Third"]=df3["Third"].replace(correct)
df3["Fourth"]=df3["Fourth"].replace(correct)

In [36]:
#Using grouby func
data_Winners = df3.groupby(["Winner"]).size().rename("Champions").reset_index()
data_Winners

Winner  Champions
0  Argentina          2
1     Brazil          5
2    England          1
3     France          1
4    Germany          4
5      Italy          4
6      Spain          1
7    Uruguay          2

In [37]:
#The following pie chart shows the countries who have won the World Cup over the years
fig=px.pie(values=data_Winners["Champions"],names=data_Winners["Winner"],height=500,width=1000)
fig.update_traces(textinfo="label+percent",textposition="inside")
fig.update_layout(title="Countries Who have Won the World Cups")
fig.show()

In [38]:
#creating some new dataframes for visualization
data_Winners = pd.DataFrame(df3["Winner"].value_counts())
data_Runner_ups = pd.DataFrame(df3["Runners-Up"].value_counts())
data_Third=pd.DataFrame(df3["Third"].value_counts())
data_Fourth=pd.DataFrame(df3["Fourth"].value_counts())

In [39]:
data_Winners

count
Winner          
Brazil         5
Italy          4
Germany        4
Uruguay        2
Argentina      2
England        1
France         1
Spain          1

In [40]:
data_Runner_ups

count
Runners-Up           
Germany             4
Argentina           3
Netherlands         3
Czechoslovakia      2
Hungary             2
Brazil              2
Italy               2
Sweden              1
France              1

In [41]:

data_Third

count
Third             
Germany          4
Brazil           2
Sweden           2
France           2
Poland           2
USA              1
Austria          1
Chile            1
Portugal         1
Italy            1
Croatia          1
Turkey           1
Netherlands      1

In [42]:
data_Fourth

count
Fourth               
Uruguay             3
Yugoslavia          2
Brazil              2
Austria             1
Sweden              1
Spain               1
Germany             1
Soviet Union        1
Italy               1
France              1
Belgium             1
England             1
Bulgaria            1
Netherlands         1
Korea Republic      1
Portugal            1

In [43]:
#Concatinatng all the data frames
countries_with_top4 = pd.concat([data_Winners,data_Runner_ups,data_Third,data_Fourth],axis=1).reset_index()
countries_with_top4.columns=["Country","Winners","Runners-Up","Third","Fourth"]
countries_with_top4.fillna(0,inplace=True)

In [44]:
#Displaying a bar graph of countries who have had the most podium finishes at the world cup
fig=px.bar(countries_with_top4,x="Country",y=["Winners","Runners-Up","Third","Fourth"],barmode="group",
           color_discrete_map={"Winners":"Gold","Runners-Up":"Silver","Third":"#CD7F32","Fourth":"Green"},width=1200,
           height=600,hover_data="value",labels={"value":"Total"},template="plotly_dark")
fig.update_traces(width=0.2)
fig.update_layout(legend=dict(x=0,y=1.02,orientation="h",title="Positions"),title="Countries With Most Podium Finishes")
fig.show()


--**Insights from the above graphs**--

From the above graph its quite clear that Brazil has won the most World cups over the years followed by Germany and Italy however in terms of the most podium finishes and being the most consistent Germany leads above the rest

In [45]:
df1_cleaned

Year              Datetime                     Stage  \
0    1930.0  13 Jul 1930 - 15:00                    Group 1   
1    1930.0  13 Jul 1930 - 15:00                    Group 4   
2    1930.0  14 Jul 1930 - 12:45                    Group 2   
3    1930.0  14 Jul 1930 - 14:50                    Group 3   
4    1930.0  15 Jul 1930 - 16:00                    Group 1   
..      ...                   ...                       ...   
847  2014.0  05 Jul 2014 - 17:00             Quarter-finals   
848  2014.0  08 Jul 2014 - 17:00                Semi-finals   
849  2014.0  09 Jul 2014 - 17:00                Semi-finals   
850  2014.0  12 Jul 2014 - 17:00   Play-off for third place   
851  2014.0  13 Jul 2014 - 16:00                      Final   

                 Stadium             City Home Team Name  Home Team Goals  \
0                Pocitos      Montevideo          France              4.0   
1         Parque Central      Montevideo             USA              3.0   
2         Parque Central      Montevideo      Yugoslavia              2.0   
3                Pocitos      Montevideo         Romania              3.0   
4         Parque Central      Montevideo       Argentina              1.0   
..                   ...              ...            ...              ...   
847     Arena Fonte Nova        Salvador     Netherlands              0.0   
848     Estadio Mineirao  Belo Horizonte          Brazil              1.0   
849   Arena de Sao Paulo       Sao Paulo     Netherlands              0.0   
850     Estadio Nacional        Brasilia          Brazil              0.0   
851  Estadio do Maracana  Rio De Janeiro         Germany              1.0   

     Away Team Goals Away Team Name                         Win conditions  \
0                1.0         Mexico                                          
1                0.0        Belgium                                          
2                1.0         Brazil                                          
3                1.0           Peru                                          
4                0.0         France                                          
..               ...            ...                                    ...   
847              0.0     Costa Rica  Netherlands win on penalties (4 - 3)    
848              7.0        Germany                                          
849              0.0      Argentina    Argentina win on penalties (2 - 4)    
850              3.0    Netherlands                                          
851              0.0      Argentina          Germany win after extra time    

     Attendance  Half-time Home Goals  Half-time Away Goals  \
0        4444.0                   3.0                   0.0   
1       18346.0                   2.0                   0.0   
2       24059.0                   2.0                   0.0   
3        2549.0                   1.0                   0.0   
4       23409.0                   0.0                   0.0   
..          ...                   ...                   ...   
847     51179.0                   0.0                   0.0   
848     58141.0                   0.0                   5.0   
849     63267.0                   0.0                   0.0   
850     68034.0                   0.0                   2.0   
851     74738.0                   0.0                   0.0   

                    Referee                  Assistant 1  \
0    LOMBARDI Domingo (URU)        CRISTOPHE Henry (BEL)   
1         MACIAS Jose (ARG)     MATEUCCI Francisco (URU)   
2       TEJADA Anibal (URU)      VALLARINO Ricardo (URU)   
3     WARNKEN Alberto (CHI)          LANGENUS Jean (BEL)   
4       REGO Gilberto (BRA)         SAUCEDO Ulises (BOL)   
..                      ...                          ...   
847   Ravshan IRMATOV (UZB)  RASULOV Abduxamidullo (UZB)   
848   RODRIGUEZ Marco (MEX)      TORRENTERA Marvin (MEX)   
849      C�neyt �AKIR (TUR)         DURAN Bahattin (TUR)   
850   HAIMOUDI Djamel (ALG)    

In [46]:
#Checking if Country names are correct
df1_without_duplicates["Home Team Name"].unique()

array(['France', 'USA', 'Yugoslavia', 'Romania', 'Argentina', 'Chile',
       'Uruguay', 'Brazil', 'Paraguay', 'Austria', 'Hungary',
       'Switzerland', 'Sweden', 'Germany', 'Spain', 'Italy',
       'Czechoslovakia', 'Cuba', 'England', 'Germany FR', 'Turkey',
       'Soviet Union', 'Northern Ireland', 'Mexico', 'Wales', 'Portugal',
       'Korea DPR', 'Peru', 'Belgium', 'Bulgaria', 'German DR', 'Zaire',
       'Poland', 'Australia', 'Scotland', 'Netherlands', 'Haiti',
       'Tunisia', 'Algeria', 'Honduras', 'Canada', 'Morocco',
       'Korea Republic', 'Iraq', 'Denmark', 'rn">United Arab Emirates',
       'Costa Rica', 'Cameroon', 'rn">Republic of Ireland', 'Colombia',
       'Norway', 'Nigeria', 'Saudi Arabia', 'Bolivia', 'Russia', 'Greece',
       'Jamaica', 'South Africa', 'Japan', 'Croatia', 'China PR',
       'Senegal', 'Slovenia', 'Ecuador', 'rn">Trinidad and Tobago',
       'rn">Serbia and Montenegro', 'Angola', 'Czech Republic', 'Togo',
       'Iran', "C�te d'Ivoire", 'Ghana

In [47]:
#Displaying all the countries with incorrect names
df1_without_duplicates[df1_without_duplicates["Home Team Name"].str.contains('rn">')]["Home Team Name"].value_counts()

Home Team Name
rn">Republic of Ireland       5
rn">United Arab Emirates      1
rn">Trinidad and Tobago       1
rn">Serbia and Montenegro     1
rn">Bosnia and Herzegovina    1
Name: count, dtype: int64

In [48]:
#Renaming them
accurate={'rn">Republic of Ireland':'Republic of Ireland','rn">United Arab Emirates':'United Arab Emirates','rn">Trinidad and Tobago':'Trinidad and Tobago',
         'rn">Serbia and Montenegro':'Serbia and Montenegro','rn">Bosnia and Herzegovina':'Bosnia and Herzegovina','German DR':'Germany','Germany FR':'Germany'}

In [49]:
#Replacing the names
df1_without_duplicates["Home Team Name"]=df1_without_duplicates["Home Team Name"].replace(accurate)
df1_without_duplicates["Away Team Name"]=df1_without_duplicates["Away Team Name"].replace(accurate)

In [50]:
#Checking
df1_without_duplicates["Home Team Name"].unique()

array(['France', 'USA', 'Yugoslavia', 'Romania', 'Argentina', 'Chile',
       'Uruguay', 'Brazil', 'Paraguay', 'Austria', 'Hungary',
       'Switzerland', 'Sweden', 'Germany', 'Spain', 'Italy',
       'Czechoslovakia', 'Cuba', 'England', 'Turkey', 'Soviet Union',
       'Northern Ireland', 'Mexico', 'Wales', 'Portugal', 'Korea DPR',
       'Peru', 'Belgium', 'Bulgaria', 'Zaire', 'Poland', 'Australia',
       'Scotland', 'Netherlands', 'Haiti', 'Tunisia', 'Algeria',
       'Honduras', 'Canada', 'Morocco', 'Korea Republic', 'Iraq',
       'Denmark', 'United Arab Emirates', 'Costa Rica', 'Cameroon',
       'Republic of Ireland', 'Colombia', 'Norway', 'Nigeria',
       'Saudi Arabia', 'Bolivia', 'Russia', 'Greece', 'Jamaica',
       'South Africa', 'Japan', 'Croatia', 'China PR', 'Senegal',
       'Slovenia', 'Ecuador', 'Trinidad and Tobago',
       'Serbia and Montenegro', 'Angola', 'Czech Republic', 'Togo',
       'Iran', "C�te d'Ivoire", 'Ghana', 'Ukraine', 'Serbia',
       'New Zealand

In [51]:
#Creating a new dataframe for visualization
df_Halftime_Home_goals =df1_without_duplicates[["Half-time Home Goals","Home Team Name","Year"]].dropna()
df_Halftime_Home_goals

Half-time Home Goals Home Team Name    Year
0                     3.0         France  1930.0
1                     2.0            USA  1930.0
2                     2.0     Yugoslavia  1930.0
3                     1.0        Romania  1930.0
4                     0.0      Argentina  1930.0
..                    ...            ...     ...
831                   1.0      Argentina  2014.0
832                   0.0    Netherlands  2014.0
833                   0.0     Costa Rica  2014.0
834                   0.0      Argentina  2014.0
835                   0.0        Belgium  2014.0

[835 rows x 3 columns]

In [52]:
#Creating a new dataframe for visualization
df_Halftime_Away_goals = df1_without_duplicates[["Half-time Away Goals","Away Team Name","Year"]].dropna()
df_Halftime_Away_goals

Half-time Away Goals Away Team Name    Year
0                     0.0         Mexico  1930.0
1                     0.0        Belgium  1930.0
2                     0.0         Brazil  1930.0
3                     0.0           Peru  1930.0
4                     0.0         France  1930.0
..                    ...            ...     ...
831                   0.0        Belgium  2014.0
832                   0.0         Mexico  2014.0
833                   0.0         Greece  2014.0
834                   0.0    Switzerland  2014.0
835                   0.0            USA  2014.0

[835 rows x 3 columns]

In [53]:
#Creating a new dataframe for visualization of top 5 countries
df_top5=df_Halftime_Home_goals[df_Halftime_Home_goals["Home Team Name"] .isin(["Brazil","Argentina","Italy","Germany","Uruguay"])]

df_top5

Half-time Home Goals Home Team Name    Year
4                     0.0      Argentina  1930.0
8                     0.0        Uruguay  1930.0
10                    3.0      Argentina  1930.0
11                    1.0         Brazil  1930.0
13                    4.0        Uruguay  1930.0
..                    ...            ...     ...
826                   0.0         Brazil  2014.0
827                   0.0         Brazil  2014.0
828                   0.0        Germany  2014.0
831                   1.0      Argentina  2014.0
834                   0.0      Argentina  2014.0

[292 rows x 3 columns]

In [54]:
#Creating a new dataframe for visualization for top 5 countries
df_top5_HG=df_top5.groupby(["Year","Home Team Name"])["Half-time Home Goals"].sum().reset_index()
df_top5_HG

Year Home Team Name  Half-time Home Goals
0   1930.0      Argentina                   6.0
1   1930.0         Brazil                   1.0
2   1930.0        Uruguay                   8.0
3   1934.0        Germany                   4.0
4   1934.0          Italy                   5.0
..     ...            ...                   ...
76  2014.0      Argentina                   2.0
77  2014.0         Brazil                   2.0
78  2014.0        Germany                   3.0
79  2014.0          Italy                   0.0
80  2014.0        Uruguay                   2.0

[81 rows x 3 columns]

In [55]:
# displaying the half time goals scored by the top 5 countries over the years at the world cup representing as the home team
fig=px.area(df_top5_HG,x="Year",y="Half-time Home Goals",facet_col="Home Team Name",facet_col_wrap=1,height=500,width=1200,
        color="Home Team Name",labels={"Half-time Home Goals":"Goals"},text="Half-time Home Goals")
fig.update_layout(title="Half-time Goals By Home Team per Year")
fig.show()

--**Insights from the above graphs**--

From the above graph we can see that while representing as a home team countries like Brazil Argentina and Germany have have been have been very attacking ,impressive and consistent as they have scored the most goals at the half time, where as teams in comparison like a grand Italy have scored lesser amount of goals at the half time while representing as a home team

In [56]:
df_top5_1=df_Halftime_Away_goals[df_Halftime_Away_goals["Away Team Name"] .isin(["Brazil","Argentina","Italy","Germany","Uruguay"])]
df_top5_1

Half-time Away Goals Away Team Name    Year
2                     0.0         Brazil  1930.0
17                    2.0      Argentina  1930.0
21                    1.0      Argentina  1934.0
23                    0.0         Brazil  1934.0
32                    0.0        Germany  1934.0
..                    ...            ...     ...
821                   0.0        Uruguay  2014.0
825                   1.0        Germany  2014.0
826                   5.0        Germany  2014.0
828                   0.0      Argentina  2014.0
829                   0.0      Argentina  2014.0

[134 rows x 3 columns]

In [57]:
#Creating a new dataframe for visualization for top 5 countries
df_top5_AG = df_top5_1.groupby(["Year","Away Team Name"])["Half-time Away Goals"].sum().reset_index()
df_top5_AG

Year Away Team Name  Half-time Away Goals
0   1930.0      Argentina                   2.0
1   1930.0         Brazil                   0.0
2   1934.0      Argentina                   1.0
3   1934.0         Brazil                   0.0
4   1934.0        Germany                   0.0
..     ...            ...                   ...
66  2014.0      Argentina                   2.0
67  2014.0         Brazil                   2.0
68  2014.0        Germany                   6.0
69  2014.0          Italy                   1.0
70  2014.0        Uruguay                   0.0

[71 rows x 3 columns]

In [58]:
# displaying the half time goals scored by the top 5 countries over the years at the world cup representing as the away team
fig=px.area(df_top5_AG,x="Year",y="Half-time Away Goals",facet_col="Away Team Name",
        facet_col_wrap=1,height=600,color="Away Team Name",labels={"Half-time Away Goals":"Goals"})
fig.update_layout(title="Half-time Goals By Away Team per Year")
fig.show()

--**Insights from the above graphs**-

From the above graph we can see that while representing as away team, teams like Germany and Brazil have scored a lot of goals at the half-time meaning they have been consistent even as an away team whereas  teams like Argentina Uruguay and Italy have scored lesser amount of goals representing as an away team

In [59]:
#USing groupby function and creating a new dataframe for visualization
df_Halftime_Home_goals_total = df_Halftime_Home_goals.groupby("Home Team Name")["Half-time Home Goals"].sum().reset_index()
df_Halftime_Home_goals_total["Home Team Name"].unique()

array(['Algeria', 'Angola', 'Argentina', 'Australia', 'Austria',
       'Belgium', 'Bolivia', 'Bosnia and Herzegovina', 'Brazil',
       'Bulgaria', 'Cameroon', 'Canada', 'Chile', 'China PR', 'Colombia',
       'Costa Rica', 'Croatia', 'Cuba', 'Czech Republic',
       'Czechoslovakia', "C�te d'Ivoire", 'Denmark', 'Ecuador', 'England',
       'France', 'Germany', 'Ghana', 'Greece', 'Haiti', 'Honduras',
       'Hungary', 'IR Iran', 'Iran', 'Iraq', 'Italy', 'Jamaica', 'Japan',
       'Korea DPR', 'Korea Republic', 'Mexico', 'Morocco', 'Netherlands',
       'New Zealand', 'Nigeria', 'Northern Ireland', 'Norway', 'Paraguay',
       'Peru', 'Poland', 'Portugal', 'Republic of Ireland', 'Romania',
       'Russia', 'Saudi Arabia', 'Scotland', 'Senegal', 'Serbia',
       'Serbia and Montenegro', 'Slovakia', 'Slovenia', 'South Africa',
       'Soviet Union', 'Spain', 'Sweden', 'Switzerland', 'Togo',
       'Trinidad and Tobago', 'Tunisia', 'Turkey', 'USA', 'Ukraine',
       'United Arab Emirates'

In [60]:
#Displaying a bar graph showing the total half time goals of all teams at the world cup while representing as the home team
fig=px.bar(df_Halftime_Home_goals_total,x="Home Team Name",y="Half-time Home Goals",width=1300,height=600,
           color="Half-time Home Goals",labels={"Half-time Home Goals":"Goals"},color_continuous_scale=px.colors.sequential.Viridis,template="plotly_dark")
fig.update_traces(width=0.5)
fig.update_layout(title="Half-time Goals By Home Teams",coloraxis_colorbar=dict(orientation="h",thickness=12,len=0.5))
fig.update_xaxes(tickangle=90)
fig.show()


--**Insights from the above graphs**-

From the above graph we can see that teams like Germany Brazil, Argentina ,Hungary and Italy have scored more goals at the half time while representing as a home team, making them and instant favourite to win a match while they are representing as the home team at the FIFA World Cup


In [61]:
df_Halftime_Away_goals_total = df_Halftime_Away_goals.groupby("Away Team Name")["Half-time Away Goals"].sum().reset_index()
df_Halftime_Away_goals_total

Away Team Name  Half-time Away Goals
0                Algeria                   4.0
1                 Angola                   0.0
2              Argentina                   8.0
3              Australia                   3.0
4                Austria                   3.0
..                   ...                   ...
76  United Arab Emirates                   1.0
77               Uruguay                   6.0
78                 Wales                   2.0
79            Yugoslavia                   8.0
80                 Zaire                   0.0

[81 rows x 2 columns]

In [62]:
#Displaying a bar graph showing the total half time goals of all teams at the world cup while representing as the away team
fig=px.bar(df_Halftime_Away_goals_total,x="Away Team Name",y="Half-time Away Goals",width=1300,height=600,
           labels={"Half-time Away Goals":"Goals"},color="Half-time Away Goals",color_continuous_scale=px.colors.sequential.Bluered,template="plotly_white")
fig.update_traces(width=0.5)
fig.update_layout(title="Half-time Goals By Away Teams",coloraxis_colorbar=dict(orientation="h",thickness=12,len=0.5))
fig.update_xaxes(tickangle=90)
fig.show()

--**Insights from the above graphs**-

The following graph shows that the total goals scored at the half time interval by countries representing as an ability from the graph we can see that Brazil , Germany , Spain , France and England has scored the most goals at the half time while representing as an away team. So in summary we can say that it doesn't really matter for countries like Brazil or Germany weather they are playing as home team or an away team as they have been somewhat consistent however it does matter for countries like Spain France England as they have been much more efficient while representing as an away team at the Half- time at the FIFA World Cup.

In [63]:
#Creating new dataframe for visualization
data_Home = df1_without_duplicates[["Home Team Goals","Home Team Name"]].dropna()
data_Home

Home Team Goals Home Team Name
0                4.0         France
1                3.0            USA
2                2.0     Yugoslavia
3                3.0        Romania
4                1.0      Argentina
..               ...            ...
831              1.0      Argentina
832              2.0    Netherlands
833              1.0     Costa Rica
834              1.0      Argentina
835              2.0        Belgium

[835 rows x 2 columns]

In [64]:
#Creating ba new dataframe for visualization
data_Away = df1_without_duplicates[["Away Team Goals","Away Team Name"]].dropna()
data_Away

Away Team Goals Away Team Name
0                1.0         Mexico
1                0.0        Belgium
2                1.0         Brazil
3                1.0           Peru
4                0.0         France
..               ...            ...
831              0.0        Belgium
832              1.0         Mexico
833              1.0         Greece
834              0.0    Switzerland
835              1.0            USA

[835 rows x 2 columns]

In [65]:
data_Home_1=data_Home.groupby("Home Team Name")["Home Team Goals"].sum().reset_index()
data_Home_1

Home Team Name  Home Team Goals
0                Algeria              5.0
1                 Angola              0.0
2              Argentina            109.0
3              Australia              7.0
4                Austria             31.0
..                   ...              ...
71  United Arab Emirates              0.0
72               Uruguay             62.0
73                 Wales              2.0
74            Yugoslavia             42.0
75                 Zaire              0.0

[76 rows x 2 columns]

In [66]:
#Dispaying a distribution of home team goals vs away team goals
hist_data=[df1_without_duplicates["Home Team Goals"],df1_without_duplicates["Away Team Goals"]]
group_labels=["Home Team Goals","Away Team Goals"]
fig=ff.create_distplot(hist_data,group_labels,colors=["purple","green"])
fig.update_layout(width=700,bargap=0.01,height=700,title="Home Team Goals Vs Away Team Goals",xaxis=dict(title="Goals"),
                  yaxis=dict(title="Density"),template="seaborn")
fig.show()

--**Insights from the above graphs**--

The following distplot shows the distributions of home goals versus away goals where we can see that the both the graphs or skewed to the right meaning they have been less matches with high number of goals and more matches with less number of goals. The peak of the distribution of home team goals is around 2 suggesting this is the common number of goal scored by the home team, where is the peak of the distribution for the away team is around 1 suggesting this is the common number of goal scored by the team.

In [67]:
#displaying a bar plot that visualises the total goals scored by countries while representing as a home team
fig=px.bar(data_Home_1,x="Home Team Name",y="Home Team Goals",width=1300,height=600,labels={"Home Team Goals":"Goals"},color="Home Team Goals"
           ,color_continuous_scale=px.colors.sequential.Jet,template="plotly_white")
fig.update_layout(title="Total Goals By Home Teams",coloraxis_colorbar=dict(orientation="h",thickness=12,len=0.5))
fig.show()

--**Insights from the above graphs**--

Above graph shows the total goal scored by countries while they are representing as home team ,from the graph it's quite clear that teams like Brazil ,Germany and Argentina are heavy favourites to win a game while representing as the home team as they have scored the most goals then the rest of the teams.

In [68]:
data_Away_1=data_Away.groupby("Away Team Name")["Away Team Goals"].sum().reset_index()
data_Away_1


Away Team Name  Away Team Goals
0                Algeria              7.0
1                 Angola              1.0
2              Argentina             22.0
3              Australia              4.0
4                Austria             12.0
..                   ...              ...
76  United Arab Emirates              2.0
77               Uruguay             18.0
78                 Wales              2.0
79            Yugoslavia             18.0
80                 Zaire              0.0

[81 rows x 2 columns]

In [69]:
#displaying a bar plot that visualises the total goals scored by countries while representing as a away team
fig=px.bar(data_Away_1,x="Away Team Name",y="Away Team Goals",width=1300,height=600,labels={"Away Team Goals":"Goals"},color="Away Team Goals"
           ,color_continuous_scale=px.colors.sequential.Purp,template="plotly_dark")
fig.update_layout(title="Total Goals By Away Teams",coloraxis_colorbar=dict(orientation="h",thickness=12,len=0.5))
fig.show()

--**Insights from graphs**--

Graph shows the total goals scored by countries while representing as an away team, from the graph, from the graph it is quiet clear that Brazil and Germany are consistent no matter which side they are representing , whereas we can see that teams like Netherlands, Spain have have been good while as an Away Team however we can also say that teams like Argentina have been less impressive as an away team as they have scored very less amount of goals as an away team then as a home team

In [70]:
#Renaming columns in dataframes
data_Home_1.columns=["Country","Goals"]
data_Away_1.columns=["Country","Goals"]

In [71]:
#Concatinating the dataframes and using grroupby func
total_goals=pd.concat([data_Home_1,data_Away_1],ignore_index=True)
total_goals1=total_goals.groupby("Country")["Goals"].sum().reset_index()
total_goals1.columns=["Country","Goals"]

In [72]:
# displaying a bar graph that visualises the total goals scored by countries at the world cup
fig=px.bar(total_goals1,x="Country",y="Goals",width=1300,height=700,text="Goals",
           color="Goals",color_continuous_scale=px.colors.sequential.Teal,template="plotly_white")
fig.update_traces(textposition="outside")
fig.update_layout(title="Total Goals By Countries At The WC",coloraxis_colorbar=dict(orientation="h",thickness=12,len=0.5))
fig.show()

--**Insights from the above graphs**--

From the graph we observe that teams like Germany and Brazil have outscored every other country at the world cup suggesting the dominance both teams have had at the world cup, it also points out the lineage of players both countries have produced over the years at the world cup like Pele, Ronaldo ,Klose and Müller just to name a few

In [73]:
#Sorting and getting the top 6 values
total_goals_sort=total_goals1.sort_values(by="Goals",ascending=False).head(6)
total_goals_sort

Country  Goals
28    Germany  227.0
8      Brazil  221.0
2   Argentina  131.0
38      Italy  128.0
27     France  106.0
67      Spain   92.0

In [74]:
#Top 6 teams in terms of goals scored at the world cup
fig=px.bar(total_goals_sort,x="Goals",y="Country",color="Country",text="Goals")
fig.update_traces(textposition="outside")
fig.update_layout(title="Top 6 Teams with Most Goals Scored")
fig.show()

In [75]:
df1_without_duplicates.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

In [76]:
df3.head()

Year      Country   Winner      Runners-Up    Third      Fourth  \
0  1930      Uruguay  Uruguay       Argentina      USA  Yugoslavia   
1  1934        Italy    Italy  Czechoslovakia  Germany     Austria   
2  1938       France    Italy         Hungary   Brazil      Sweden   
3  1950       Brazil  Uruguay          Brazil   Sweden       Spain   
4  1954  Switzerland  Germany         Hungary  Austria     Uruguay   

   GoalsScored  QualifiedTeams  MatchesPlayed Attendance  
0           70              13             18    590.549  
1           70              16             17    363.000  
2           84              15             18    375.700  
3           88              13             22  1.045.246  
4          140              16             26    768.607

In [77]:
df2_without_duplicates

RoundID    MatchID Team Initials               Coach Name Line-up  \
0          201       1096           FRA      CAUDRON Raoul (FRA)       S   
1          201       1096           MEX         LUQUE Juan (MEX)       S   
2          201       1096           FRA      CAUDRON Raoul (FRA)       S   
3          201       1096           MEX         LUQUE Juan (MEX)       S   
4          201       1096           FRA      CAUDRON Raoul (FRA)       S   
...        ...        ...           ...                      ...     ...   
37043   255951  300186497           USA  KLINSMANN Juergen (GER)       N   
37044   255951  300186497           BEL       WILMOTS Marc (BEL)       N   
37045   255951  300186497           USA  KLINSMANN Juergen (GER)       N   
37046   255951  300186497           BEL       WILMOTS Marc (BEL)       N   
37047   255951  300186497           USA  KLINSMANN Juergen (GER)       N   

       Shirt Number       Player Name Position Event  
0                 0       Alex THEPOT       GK   NaN  
1                 0   Oscar BONFIGLIO       GK   NaN  
2                 0  Marcel LANGILLER      NaN  G40'  
3                 0      Juan CARRENO      NaN  G70'  
4                 0   Ernest LIBERATI      NaN   NaN  
...             ...               ...      ...   ...  
37043            17          ALTIDORE      NaN   NaN  
37044            16            DEFOUR      NaN   NaN  
37045            18       WONDOLOWSKI      NaN  I72'  
37046            21      VANDEN BORRE      NaN   NaN  
37047            21          CHANDLER      NaN   NaN  

[37048 rows x 9 columns]

In [78]:
#Creating a new dataframe for visualization
data_Coach = df2_without_duplicates.groupby("Coach Name")["MatchID"].nunique().reset_index()

In [79]:
#Sorting the dataframe
coach_sort=data_Coach.sort_values(by="MatchID",ascending=False)

In [80]:
top_6=coach_sort.head(6)

In [81]:
top_6.columns=["Coach Name","Total Matches"]
top_6

Coach Name  Total Matches
277            SCHOEN Helmut (FRG)             25
222  PARREIRA Carlos Alberto (BRA)             23
278      SCOLARI Luiz Felipe (BRA)             21
333            ZAGALLO Mario (BRA)             20
189         MILUTINOVIC Bora (YUG)             20
18              BEARZOT Enzo (ITA)             18

In [82]:
#Displaying the top 6 coaches who have managed the most games at the World cup
fig=px.bar(top_6,x="Total Matches",y="Coach Name",color="Coach Name",text="Total Matches",color_discrete_sequence=px.colors.sequential.Blackbody)
fig.update_traces(textposition="outside")
fig.update_layout(title="Coaches With Most World Cup Appearances")
fig.show()

--**Insights from the above graphs**--

From the above graph its quite clear that these are the these are the coaches that have managed the most games at the FIFA World Cup, graph also suggest that these managers have managed over multiple World cup campaigns,it maybe due the promise they have shown or achievements they have had at the FIFA world cup

In [83]:
#Total no. of players who have featured at the World Cup
no_of_players = df2_without_duplicates["Player Name"].nunique()

In [84]:
df1_without_duplicates.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  Attendance  Half-time Home Goals  Half-time Away Goals  \
0                     4444.0                   3.0                   0.0   
1                    18346.0                   2.0                   0.0   
2                    24059.0                   2.0                   0.0   
3                     2549.0                   1.0                   0.0   
4                    23409.0                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2  RoundID  MatchID Home Team Initials  \
0         REGO Gilberto (BRA)    201.0   1096.0                FRA   
1       WARNKEN Alberto (CHI)    201.0   1090.0                USA   
2         BALWAY Thomas (FRA)    201.0   1093.0                YUG   
3    MATEUCCI Francisco (URU)    201.0   1098.0                ROU   
4  RADULESCU Constantin (ROU)    201.0   1085.0                ARG   

  Away Team Initials  
0                MEX  
1                BEL  
2                BRA  
3                PER  
4                FRA

In [85]:
#Checking out the results at the world cup
def Winner(df1_without_duplicates):
    if df1_without_duplicates['Home Team Goals'] > df1_without_duplicates['Away Team Goals']:
        return 'Home team won'
    elif df1_without_duplicates['Home Team Goals'] < df1_without_duplicates['Away Team Goals']:
        return 'Away Team won'
    elif df1_without_duplicates['Home Team Goals'] == df1_without_duplicates['Away Team Goals']:
        return 'Draw'

In [86]:
#Creating a new column for the results
df1_without_duplicates['Winner']=df1_without_duplicates.apply(lambda x:Winner(x),axis=1)
df1_without_duplicates.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Home Goals  Half-time Away Goals  \
0                 ...                   3.0                   0.0   
1                 ...                   2.0                   0.0   
2                 ...                   2.0                   0.0   
3                 ...                   1.0                   0.0   
4                 ...                   0.0                   0.0   

                  Referee               Assistant 1  \
0  LOMBARDI Domingo (URU)     CRISTOPHE Henry (BEL)   
1       MACIAS Jose (ARG)  MATEUCCI Francisco (URU)   
2     TEJADA Anibal (URU)   VALLARINO Ricardo (URU)   
3   WARNKEN Alberto (CHI)       LANGENUS Jean (BEL)   
4     REGO Gilberto (BRA)      SAUCEDO Ulises (BOL)   

                  Assistant 2 RoundID  MatchID  Home Team Initials  \
0         REGO Gilberto (BRA)   201.0   1096.0                 FRA   
1       WARNKEN Alberto (CHI)   201.0   1090.0                 USA   
2         BALWAY Thomas (FRA)   201.0   1093.0                 YUG   
3    MATEUCCI Francisco (URU)   201.0   1098.0                 ROU   
4  RADULESCU Constantin (ROU)   201.0   1085.0                 ARG   

  Away Team Initials         Winner  
0                MEX  Home team won  
1                BEL  Home team won  
2                BRA  Home team won  
3                PER  Home team won  
4                FRA  Home team won  

[5 rows x 21 columns]

In [87]:
#Results
df1_without_duplicates["Winner"].value_counts()

Winner
Home team won    478
Draw             186
Away Team won    171
Name: count, dtype: int64

In [88]:
#Donut chart showing the results at the world cup
fig=px.pie(df1_without_duplicates,names='Winner',labels="Winners",hole=0.5,color_discrete_sequence=px.colors.cyclical.Edge,width=800,height=400)
fig.update_layout(title="Results over the World Cup")
fig.update_traces(textinfo="label+percent")
fig.show()

In [89]:
#Results of matches that ended as draw at full time
df1_without_duplicates["Win conditions"].nunique()

43

--**Insights from the above graphs**-

The following donut chart show the results of all the games at the world cup where we can see that being home team has been really favourable in order to win a game as teams representing the Home Team have won the most amount of games where as being an Away Team has lesser chances of winning the game. We have also seen that there have been many draws at the world cup however out of those 186 draws 43 matches have been settled either at the extra time or at the penalties

In [90]:
#Total no of stadiums used at the world cup
no_of_staiums_used= df1_without_duplicates["Stadium"].nunique()


In [91]:
#Merging the datframes for visualization
df1_without_duplicates=df1_without_duplicates.merge(df3[["Year","Country"]],on="Year",how="left")
df1_without_duplicates.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Away Goals                 Referee  \
0                 ...                   0.0  LOMBARDI Domingo (URU)   
1                 ...                   0.0       MACIAS Jose (ARG)   
2                 ...                   0.0     TEJADA Anibal (URU)   
3                 ...                   0.0   WARNKEN Alberto (CHI)   
4                 ...                   0.0     REGO Gilberto (BRA)   

                Assistant 1                 Assistant 2 RoundID MatchID  \
0     CRISTOPHE Henry (BEL)         REGO Gilberto (BRA)   201.0  1096.0   
1  MATEUCCI Francisco (URU)       WARNKEN Alberto (CHI)   201.0  1090.0   
2   VALLARINO Ricardo (URU)         BALWAY Thomas (FRA)   201.0  1093.0   
3       LANGENUS Jean (BEL)    MATEUCCI Francisco (URU)   201.0  1098.0   
4      SAUCEDO Ulises (BOL)  RADULESCU Constantin (ROU)   201.0  1085.0   

   Home Team Initials  Away Team Initials         Winner  Country  
0                 FRA                 MEX  Home team won  Uruguay  
1                 USA                 BEL  Home team won  Uruguay  
2                 YUG                 BRA  Home team won  Uruguay  
3                 ROU                 PER  Home team won  Uruguay  
4                 ARG                 FRA  Home team won  Uruguay  

[5 rows x 22 columns]

In [92]:
#Renaming Columns
df1_without_duplicates.rename(columns={"Country_x":"Country"},inplace=True)

In [93]:
df1_without_duplicates.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Away Goals                 Referee  \
0                 ...                   0.0  LOMBARDI Domingo (URU)   
1                 ...                   0.0       MACIAS Jose (ARG)   
2                 ...                   0.0     TEJADA Anibal (URU)   
3                 ...                   0.0   WARNKEN Alberto (CHI)   
4                 ...                   0.0     REGO Gilberto (BRA)   

                Assistant 1                 Assistant 2 RoundID MatchID  \
0     CRISTOPHE Henry (BEL)         REGO Gilberto (BRA)   201.0  1096.0   
1  MATEUCCI Francisco (URU)       WARNKEN Alberto (CHI)   201.0  1090.0   
2   VALLARINO Ricardo (URU)         BALWAY Thomas (FRA)   201.0  1093.0   
3       LANGENUS Jean (BEL)    MATEUCCI Francisco (URU)   201.0  1098.0   
4      SAUCEDO Ulises (BOL)  RADULESCU Constantin (ROU)   201.0  1085.0   

   Home Team Initials  Away Team Initials         Winner  Country  
0                 FRA                 MEX  Home team won  Uruguay  
1                 USA                 BEL  Home team won  Uruguay  
2                 YUG                 BRA  Home team won  Uruguay  
3                 ROU                 PER  Home team won  Uruguay  
4                 ARG                 FRA  Home team won  Uruguay  

[5 rows x 22 columns]

In [94]:
#Using groupby and renaming columns
most_stadiums = df1_without_duplicates.groupby("Country")["Stadium"].nunique().reset_index()
most_stadiums.columns=["Country","Total Stadiums"]

In [95]:
#displaying a scatter plot to visualise the different number of stadiums used at different countries over the FIFA World Cup
fig=px.scatter(most_stadiums,x="Country",y="Total Stadiums",color="Country",text="Total Stadiums",size="Total Stadiums",color_discrete_sequence=px.colors.sequential.Rainbow)
fig.update_traces(textposition="top center")
fig.update_layout(title="Total Stadiums Used per Country in the World Cup")
fig.show()

--**Insights from the above graphs**-

From the above scatter plot its clear that Korea/Japan and Italy and Brazil have had the most number of stadiums used at the World cups giving the fans a varied and unique atmosphere throughout the world cup

In [96]:
#No of players representing their countries
total_players_per_country=df2_without_duplicates.groupby("Team Initials")["Player Name"].count().reset_index()
total_players_per_country.columns=["Team Initials","Total Players"]
total_players_per_country


Team Initials  Total Players
0            ALG            293
1            ANG             69
2            ARG           1715
3            AUS            296
4            AUT            638
..           ...            ...
77           URU           1136
78           USA            706
79           WAL            110
80           YUG            799
81           ZAI             66

[82 rows x 2 columns]

In [97]:
#Sorting ang getting top 10 values
countries_with_most_players=total_players_per_country.sort_values(by="Total Players",ascending=False).head(10)
countries_with_most_players

Team Initials  Total Players
8            BRA           2311
39           ITA           1843
2            ARG           1715
23           ENG           1378
26           FRG           1364
24           ESP           1317
25           FRA           1298
46           MEX           1167
77           URU           1136
47           NED           1118

In [98]:
# displaying a bar plot to visualise the top10 teams with most number of player representations at the FIFA World Cup
fig=px.bar(countries_with_most_players,x="Team Initials",y="Total Players",color="Total Players",color_continuous_scale=px.colors.diverging.Earth,text="Total Players")
fig.update_layout(title="Teams with Most Representations At The World Cup ")
fig.show()

--**Insights from the above graphs**--

From the above graph it is quiet clear that Brazil has had the most representations at the world cup with the number of players being around 2300 followed by Italy with around 1800 players

In [99]:
#Checking the different categorical values in Stage Column
df1_without_duplicates["Stage"].unique()

array(['Group 1', 'Group 4', 'Group 2', 'Group 3', 'Semi-finals', 'Final',
       'Preliminary round', 'Quarter-finals', 'Match for third place',
       'First round', 'Group 6', 'Group B', 'Group A', 'Group 5',
       'Group C', 'Group D', 'Group F', 'Group E', 'Round of 16',
       'Group H', 'Group G', 'Third place', 'Play-off for third place'],
      dtype=object)

In [100]:
#Filtering by Finals only
df1_finals=df1_without_duplicates[df1_without_duplicates["Stage"]=="Final"]

In [101]:
df1_finals.head(5)

Year              Datetime  Stage             Stadium         City  \
17   1930.0  30 Jul 1930 - 14:15   Final  Estadio Centenario  Montevideo    
34   1934.0  10 Jun 1934 - 17:30   Final       Nazionale PNF        Rome    
52   1938.0  19 Jun 1938 - 17:00   Final     Stade Olympique    Colombes    
100  1954.0  04 Jul 1954 - 17:00   Final    Wankdorf Stadium       Berne    
135  1958.0  29 Jun 1958 - 15:00   Final     Rasunda Stadium       Solna    

    Home Team Name  Home Team Goals  Away Team Goals  Away Team Name  \
17         Uruguay              4.0              2.0       Argentina   
34           Italy              2.0              1.0  Czechoslovakia   
52           Italy              4.0              2.0         Hungary   
100        Germany              3.0              2.0         Hungary   
135         Brazil              5.0              2.0          Sweden   

                  Win conditions  ...  Half-time Away Goals  \
17                                ...                   2.0   
34   Italy win after extra time   ...                   0.0   
52                                ...                   1.0   
100                               ...                   2.0   
135                               ...                   1.0   

                      Referee               Assistant 1  \
17        LANGENUS Jean (BEL)      SAUCEDO Ulises (BOL)   
34          EKLIND Ivan (SWE)         BAERT Louis (BEL)   
52   CAPDEVILLE Georges (FRA)      WUETHRICH Hans (SUI)   
100        LING William (ENG)  ORLANDINI Vincenzo (ITA)   
135      GUIGUE Maurice (FRA)        DUSCH Albert (GER)   

                  Assistant 2 RoundID MatchID  Home Team Initials  \
17      CRISTOPHE Henry (BEL)   405.0  1087.0                 URU   
34     IVANCSICS Mihaly (HUN)  3490.0  1134.0                 ITA   
52         KRIST Gustav (TCH)  3487.0  1174.0                 ITA   
100  GRIFFITHS Benjamin (WAL)  3484.0  1278.0                 FRG   
135    GARDEAZABAL Juan (ESP)  3482.0  1343.0                 BRA   

     Away Team Initials         Winner      Country  
17                  ARG  Home team won      Uruguay  
34                  TCH  Home team won        Italy  
52                  HUN  Home team won       France  
100                 HUN  Home team won  Switzerland  
135                 SWE  Home team won       Sweden  

[5 rows x 22 columns]

In [102]:
#Countries with most finals played as away team
top_most_finals1=df1_finals.groupby(["Home Team Name"])["Stage"].count().reset_index()
top_most_finals1.columns=["Country","Finals"]
top_most_finals1

Country  Finals
0    Argentina       2
1       Brazil       5
2      England       1
3      Germany       4
4        Italy       4
5  Netherlands       2
6      Uruguay       1

In [103]:
#Countries with most finals played as away team
top_most_finals2=df1_finals.groupby(["Away Team Name"])["Stage"].count().reset_index()
top_most_finals2.columns=["Country","Finals"]
top_most_finals2

Country  Finals
0       Argentina       3
1          Brazil       1
2  Czechoslovakia       2
3          France       2
4         Germany       4
5         Hungary       2
6           Italy       2
7     Netherlands       1
8           Spain       1
9          Sweden       1

In [104]:
#Concating both dataframes
top_most_finals_total = pd.concat([top_most_finals1,top_most_finals2])
top_most_finals_total

Country  Finals
0       Argentina       2
1          Brazil       5
2         England       1
3         Germany       4
4           Italy       4
5     Netherlands       2
6         Uruguay       1
0       Argentina       3
1          Brazil       1
2  Czechoslovakia       2
3          France       2
4         Germany       4
5         Hungary       2
6           Italy       2
7     Netherlands       1
8           Spain       1
9          Sweden       1

In [105]:
#Using groupby function
top_most_finals_total1=top_most_finals_total.groupby("Country")["Finals"].sum().reset_index()

In [106]:
#Displaying a barplot of countries who have playede the most finals
fig= px.bar(top_most_finals_total1,x="Finals",y="Country",color="Country",text="Finals",template="plotly_dark")
fig.update_traces(textposition="outside")
fig.update_layout(title="Countries with Most Finals Played")
fig.show()

--**Insights from the above graphs**--

From the above graph we can see that Germany followed by Brazil and Italy have played the most finals out all the other countries and thus probably hinting out why Germany and Brazil had outscored all the other teams in the world cup

In [107]:
#Renaming a value used in the visualization
correct1={'Maracan� - Est�dio Jornalista M�rio Filho':'Maracanã - Estádio Jornalista Mário Filho'}

In [108]:
#Replacing with renamed value
df1_without_duplicates["Stadium"]=df1_without_duplicates["Stadium"].replace(correct1)

In [109]:
#Using groupby func
stadiums_with_most_avg_attendance=df1_without_duplicates.groupby("Stadium")["Attendance"].mean().reset_index()

In [110]:
#Sorting out
stadiums_with_most_avg_attendance_sort=stadiums_with_most_avg_attendance.sort_values(by="Attendance",ascending=False).head(6)

In [111]:
#Displaying a bar plot for the top six stadiums with the highest average attendances at the FIFA World Cup
fig= px.bar(stadiums_with_most_avg_attendance_sort,x="Attendance",y="Stadium",color="Stadium",text="Attendance",height=600,color_discrete_sequence=px.colors.sequential.Electric)
fig.update_traces(textposition="inside")
fig.update_layout(title="Stadiums with Highest Average Attendance")
fig.show()

--**Insights from the above graphs**--

From the above bar plot we can see that the Marcanā stadium has had the highest average attendance for all the World cups as it infamously held the 2014 FIFA world cup final which was won by Germany

In [112]:
df1_without_duplicates.head()

Year              Datetime    Stage         Stadium         City  \
0  1930.0  13 Jul 1930 - 15:00   Group 1         Pocitos  Montevideo    
1  1930.0  13 Jul 1930 - 15:00   Group 4  Parque Central  Montevideo    
2  1930.0  14 Jul 1930 - 12:45   Group 2  Parque Central  Montevideo    
3  1930.0  14 Jul 1930 - 14:50   Group 3         Pocitos  Montevideo    
4  1930.0  15 Jul 1930 - 16:00   Group 1  Parque Central  Montevideo    

  Home Team Name  Home Team Goals  Away Team Goals Away Team Name  \
0         France              4.0              1.0         Mexico   
1            USA              3.0              0.0        Belgium   
2     Yugoslavia              2.0              1.0         Brazil   
3        Romania              3.0              1.0           Peru   
4      Argentina              1.0              0.0         France   

  Win conditions  ...  Half-time Away Goals                 Referee  \
0                 ...                   0.0  LOMBARDI Domingo (URU)   
1                 ...                   0.0       MACIAS Jose (ARG)   
2                 ...                   0.0     TEJADA Anibal (URU)   
3                 ...                   0.0   WARNKEN Alberto (CHI)   
4                 ...                   0.0     REGO Gilberto (BRA)   

                Assistant 1                 Assistant 2 RoundID MatchID  \
0     CRISTOPHE Henry (BEL)         REGO Gilberto (BRA)   201.0  1096.0   
1  MATEUCCI Francisco (URU)       WARNKEN Alberto (CHI)   201.0  1090.0   
2   VALLARINO Ricardo (URU)         BALWAY Thomas (FRA)   201.0  1093.0   
3       LANGENUS Jean (BEL)    MATEUCCI Francisco (URU)   201.0  1098.0   
4      SAUCEDO Ulises (BOL)  RADULESCU Constantin (ROU)   201.0  1085.0   

   Home Team Initials  Away Team Initials         Winner  Country  
0                 FRA                 MEX  Home team won  Uruguay  
1                 USA                 BEL  Home team won  Uruguay  
2                 YUG                 BRA  Home team won  Uruguay  
3                 ROU                 PER  Home team won  Uruguay  
4                 ARG                 FRA  Home team won  Uruguay  

[5 rows x 22 columns]

In [113]:
#Referees with most matches officiated at the world cup
df_Refs= df1_without_duplicates.groupby("Referee")["MatchID"].nunique().reset_index()
df_Refs

Referee  MatchID
0           ABD EL FATAH Essam (EGY)        1
1                   ADAIR John (NIR)        1
2                AGNOLIN Luigi (ITA)        4
3        AGUILAR ELIZALDE Abel (MEX)        1
4                 AGUILAR Joel (SLV)        2
..                               ...      ...
361              WYSSLING Paul (SUI)        4
362             Wolfgang STARK (GER)        3
363  YAMASAKI MALDONADO Arturo (MEX)        5
364         ZECEVIC Konstantin (YUG)        1
365               ZSOLT Istvan (HUN)        5

[366 rows x 2 columns]

In [114]:
df_Refs_sort=df_Refs.sort_values(by="MatchID",ascending=False).head(5)
df_Refs_sort=df_Refs_sort.rename(columns={"MatchID":"Total Matches"})

In [115]:
#Displaying a bar plot for top 5 referees who have officiated the most games
fig= px.bar(df_Refs_sort,x="Total Matches",y="Referee",color="Referee",text="Total Matches",height=600
,color_discrete_sequence=px.colors.sequential.Turbo,template="plotly_dark")
fig.update_traces(textposition="inside")
fig.update_layout(title="Referees With Most Games Officiated At The World Cup")
fig.show()